compiler/pipes/final-check.cpp

```diff
@@ -408,10 +411,50 @@ VertexPtr FinalCheckPass::on_enter_vertex(VertexPtr vertex) {
     check_null_usage_in_binary_operations(binary_vertex);
   }
 
+  if (vertex->type() == op_index) {
+    check_indexing(vertex);
+  }
+
   //TODO: may be this should be moved to tinf_check
   return vertex;
 }
 
+void FinalCheckPass::check_indexing(VertexPtr vertex) {
+  const auto array = vertex.as<op_index>()->array();
+  const auto key = vertex.as<op_index>()->key();
+  const auto *key_type = tinf::get_type(key);
+  const auto *array_type = tinf::get_type(array);
+  if (key_type == nullptr || array_type == nullptr) {
+    return;
+  }
+
+  const auto key_ptype = key_type->ptype();
+  const auto key_type_is_raw = !key_type->or_false_flag() && !key_type->or_null_flag();
+  bool is_allowed = false;
+  std::string allowed_types;
+  std::string what_indexing;
+
+  switch (array_type->ptype()) {
+    case tp_tuple:
+    case tp_string:
```

Comment:
tp_shape?

Reply:
Are we not allowing strings when indexing shapes?

Reply:
you have the special logic for tp_tuple, but you don't have it for tp_shape is it ok?

Reply:
Since the validation logic for shapes is the same as validation for all other types, does it make sense to put this in a separate case?

Reply:
ouch, I see, ignore this comment